### DISCUSSION TITLE
Inversion with custom forward operators - custom objective function and forward model definition

### DISCUSSION THREAD

**Discussion by LuFrag** [gimli-org/pyGIMLi#494] [Q&A] [answered]
Hello everyone,

I created a custom forward operator for a GPR propagation in 1-D.

I took a look at the tutorial @ Transform 2021 "Inversion with custom forward operators" and I have two questions.

1. Can I change the shape and the definition of the misfit (objective) function ? I would like to normalise the misfit by the norm of the mean value of my GPR waveform instead of the error estimation.

2. In the tutorial example, the forward model class is defined as the code below. The 'response' method can accept other arguments besides the parameter model ? In my case, I have parameters like time window or the source signal that must be passed to the function "fwd", but they will not be inverted. They can be passed as *kwargs or they have to be defined outside the function like in the empymod exemple ?

```
class myFwd(pg.Modelling):
    def __init__(self, depth):
        """Initialize the model."""
        self.dep = depth
        self.mesh1d = pg.meshtools.createMesh1D(len(self.dep))
        super().__init__()
        self.setMesh(self.mesh1d)
    
    def response(self, model):
        """Forward response."""
        A = fwd(model, self.dep)
        Avec = A.ravel()
        return np.hstack((Avec.real, Avec.imag))
        
    def createStartModel(self):
        return pg.Vector(len(self.dep)) * 100
```

Thanks again, please let me know if you need further information.

Luis

**Comment by halbmy** [comment, marked as answer]
1. I am afraid not. The data objective function is rather fixed with the error-weighted misfit norm, but of course you can adapt the error accordingly.
2. The response function just takes the model (which can be a parameter distribution in any dimension but also additional parameters you want to invert) as a vector, this is the basic algebraic definition. Any additional parameters (that are not to be inverted) should be stored in the class (`self.myVeryImportantParameter`) when initializing the fop. Any parameter that should be part of inversion needs to be stripped from the vector and can be defined by an additional region with own value limits, constraints etc.

**Comment by LuFrag** [comment]
Thanks for your answers.

1. Ok. If I understand correctly, this means that if define a relativeError vector of 100% (1 in this case), the original chi² objective function will be transformed in a squared relative RMS objective function ? (RRMS as defined in https://github.com/gimli-org/gimli/discussions/442).

2. I come up with the following "general" code for a 1-D inversion based on your answer. Is it on of the ways to use self.myVeryImportantParameter ?


```
# Forward model function
def fwd(par,depth):
    
    A = ForwardModel(par,depth)
    
    return A
    
# Class
class myFwd(pg.Modelling):
    def __init__(self, depth, Important_not_invert):
        """Initialize the model."""
        self.dep = depth
        self.mesh1d = pg.meshtools.createMesh1D(len(self.dep))
        super().__init__()
        self.setMesh(self.mesh1d)
        
        self.MyVeryImportantParameter = Important_not_invert
    
    def response(self, model):
        """Forward response."""
        
        par = model[0:2] # Strip the original vector
        
        res = fwd(par, self.dep) # Only associate the parameter to be inverted
       
        return res
        
    def createStartModel(self):
        return pg.Vector(len(self.dep)+1) * np.log(12)
    
#Calling the functions    
model = [par_to_invert1,par_to_invert2,Important_not_invert]

depth = [1,2,4]

fop = myFwd(depth,model[2])

fop.response(model)
```

**Reply by halbmy** [reply]
Yes, this looks very good!

**Comment by halbmy** [comment]
Any news on this or can we close the issue?
